documentation/specs/Stepper.md

```diff
@@ -0,0 +1,102 @@
+- The step sequence is non-linear, which means users can navigate to previously completed steps.
+- On small screens, the step label is not visible.
+- A step that is not completed or active is considered disabled.
+
+---
+
+## Design
+
+In contrast to most Easy UI components, the `Stepper` component does not have an immediate `react-aria` primitive to build on top of.
+
+The `Stepper` component has two considerations: The outer container and individual steps. The outer container will handle orientation and spacing concerns with each individual step with respect to another. Individual steps will be represented as `Stepper.Step` and each `Stepper.Step` renders a `Stepper.StepButton` and behind the scenes a `Stepper.StepSeparator`. The `Stepper.StepButton` will make use of the `UnstyledButton` component and the clickable area will not only include the label, but a left aligned icon as well. Representing each step as a button is necessary to support a non-linear step sequence where previously completed steps can be clicked and returned to. The `Stepper.StepSeparator` will not be exposed to the consumer, and will be omitted for the last step in the collection.
+
+### API
+
+```ts
+type StepperColor = "primary" | "inverse";
+type StepperOrientation = "horizontal" | "vertical";
+
+type Stepper = {
+  /**
+   * Color scheme that applies to stepper in various states.
+   */
+  color?: StepperColor;
+  /**
+   * Orientation of stepper, vertical replaces arrow icon with line.
+   */
+  orientation?: StepperOrientation;
+  /**
+   * The current step, other steps are either in a disabled or completed state.
+   */
+  activeStep: number;
+  /**
+   * The steps to render, individual Step components.
+   */
+  children: ReactNode;
+};
+
+type Step = {
+  /**
+   * Whether or not the step should have completed styles.
+   */
+  isCompleted: boolean;
+  /**
+   * Renders StepButton and StepSeparator.
+   */
+  children: ReactNode;
+};
+
+type StepButton = AriaButtonProps & {
+  /**
+   * StepButton content, the actual step text itself.
+   */
+  children: ReactNode;
+};
+```
+
+### Example Usage
+
+_Basic usage_:
+
+```tsx
+import { useState } from "react";
+import { Stepper } from "@easypost/easy-ui/Stepper";
+
+const steps = ["Step 1", "Step 2", "Step 3"];
+
+function App() {
+  const [activeStep, setActiveStep] = useState(0);
+  const [completed, setCompleted] = React.useState<{
+    [index: number]: boolean;
+  }>({});
+
+  const handleStep = (index) => setActiveStep(index);
+
+  return (
+    <Stepper color="primary" orientation="horizontal" activeStep={activeStep}>
+      {steps.map((stepLabel, index) => (
+        <Stepper.Step key={stepLabel} isCompleted={status[index]}>
+          <Stepper.StepButton onPress={handleStep(index)}>
+            {stepLabel}
+          </Stepper.StepButton>
+        </Stepper.Step>
+      ))}
+    </Stepper>
+  );
+}
+```
```

Comment:
i wouldn't say i love this design and it is certainly subject to change during the implementation phase. other approaches considered was to have a `<Stepper />` component that accepts a "data" object similar to what exists on the site currently; in other words omitting `Stepper.Step` and `Stepper.StepButton` and everything is handled behind the scenes. maybe thats a bit easier for consumers to reason about too.

Reply:
i like the thought put into this!

the design makes sense coming from Chakra and MUI. i think for our use case, imo, we can tighten up the API a little, since we don't have to accommodate the same kind of flexibility that the general purpose Chakra/MUI libraries have to accommodate. they're basically giving consumers building blocks that they can use as their needs require.

for us, i think we can go simpler and just have a `Stepper` and `Stepper.Step` components, basically locking down the styling/flexibility to only suit our needs. i like having those two components instead of a `data` prop personally, as i think it keeps to the spirit of the rest of our component library and provides a nice parallel in the React tree with the visuals on the screen.

inside the `Stepper.Step` component we can have any of the implementation details, like whether to render a button or separator, etc.

Reply:
good point. i think id like to see how the implementation shakes out, but i think it should be feasible to just have `Stepper` and `Stepper.Step`